engine/state_manager.go

```diff
@@ -0,0 +1,136 @@
+package engine
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/thrasher-corp/gocryptotrader/common"
+	exchange "github.com/thrasher-corp/gocryptotrader/exchanges"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/asset"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/state"
+	"github.com/thrasher-corp/gocryptotrader/log"
+)
+
+const (
+	StateManagement          = "StateManagement"
+	DefaultStateManagerDelay = time.Minute
+)
+
+var (
+	errSubsystemNotSetup = errors.New("subsystem not set up")
+	errNilInterface      = errors.New("interface is nil")
+)
+
+// stateManager provides external package coupling for method application
+type stateManager struct {
+	started  int32
+	shutdown chan struct{}
+	wg       sync.WaitGroup
+	*state.Manager
+	iExchangeManager
+	sleep time.Duration
+}
+
+// Setup applies configuration parameters before running
+func (s *stateManager) Setup(interval time.Duration, em iExchangeManager) error {
+	if em == nil {
+		return errNilInterface
+	}
+	s.sleep = interval
+	if interval <= 0 {
+		log.Warnf(log.ExchangeSys,
+			"%s interval is invalid, defaulting to: %s",
+			StateManagement,
+			DefaultStateManagerDelay)
+		s.sleep = DefaultStateManagerDelay
+	}
+	s.Manager = state.GetManager()
+	s.iExchangeManager = em
+	s.shutdown = make(chan struct{})
+	return nil
+}
+
+// Start runs the subsystem
+func (s *stateManager) Start() error {
+	if s == nil {
+		return fmt.Errorf("%s %w", StateManagement, ErrNilSubsystem)
+	}
+
+	if s.Manager == nil {
+		return fmt.Errorf("%s %w", StateManagement, errSubsystemNotSetup)
+	}
+
+	if !atomic.CompareAndSwapInt32(&s.started, 0, 1) {
+		return fmt.Errorf("%s %w", StateManagement, ErrSubSystemAlreadyStarted)
+	}
+	s.wg.Add(1)
+	go s.monitor()
+	return nil
+}
+
+// Stop stops the subsystem
+func (s *stateManager) Stop() error {
+	if s == nil {
+		return fmt.Errorf("%s %w", StateManagement, ErrNilSubsystem)
+	}
+	if atomic.LoadInt32(&s.started) == 0 {
+		return fmt.Errorf("%s %w", StateManagement, ErrSubSystemNotStarted)
+	}
+
+	log.Debugf(log.ExchangeSys, "%s %s", StateManagement, MsgSubSystemShuttingDown)
+	close(s.shutdown)
+	s.wg.Wait()
+	s.shutdown = make(chan struct{})
+	log.Debugf(log.ExchangeSys, "%s %s", StateManagement, MsgSubSystemShutdown)
+	atomic.CompareAndSwapInt32(&s.started, 1, 0)
```

Comment:
Can be changed to `StoreInt32` for performance as you've already done a value check at start. Its 2 whole nanoseconds in savings! :D Unless its because you're worried that it could fail from another status change elsewhere?